teps/0142-enable-step-reusability.md

```diff
@@ -0,0 +1,1089 @@
+# declare the params that it needs.
+apiVersion: tekton.dev/v1alpha1
+kind: StepAction
+metadata:
+  name: myStep
+spec:
+  params:
+    - name: url
+      description: “This is a param”
+      default: bar
+  image: alpine
+  script: |
+    echo $(params.url)
+
+```
+</td>
+<td rowspan=2>
+
+```yaml
+# This is how a task author would provide 
+# the params that the step needs.
+
+apiVersion: tekton.dev/v1
+kind: Task
+metadata:
+  name: ci-t
+spec:
+  params:
+    - name: url1
+    - name: url2
+  steps:
+    - name: step1
+      ref: 
+        name: myStep
+      params:
+        - name: url
+          value: url1
+    - name: step2
+      ref:
+        name: myOtherStep
+      params:
+        - name: url
+          value: url2
+```
+</td>
+</tr>
+<tr>
+<td>
+
+```yaml
+apiVersion: tekton.dev/v1alpha1
+kind: StepAction
+metadata:
+  name: myStep
+spec:
+  params:
+    - name: url
+      description: “Another param”
+      default: baz
+  image: alpine
+  script: |
+    echo $(params.url)
+
+```
+</td>
+</tr>
+</table>
+
+#### Results
+
+A referenced `StepAction` needs to declare if it produces a `Result` so that `Task` authors know that they can reference it without having to drill into the `StepAction` specification. 
+
+If multiple `StepActions` produce the different `Results` under the same name then to ensure that they are not overwritten, the `Task` needs to be able to capture them under different names. In order to capture these at the task level, we need to: 
+1. Introduce a field to [TaskResults](https://github.com/tektoncd/pipeline/blob/445734d92807a80158b4b7af605d768c647fdb3d/pkg/apis/pipeline/v1/result_types.go#L19-L49) called `Value` like we do for [PipelineResults](https://github.com/tektoncd/pipeline/blob/445734d92807a80158b4b7af605d768c647fdb3d/pkg/apis/pipeline/v1/pipeline_types.go#L129C1-L144C2). (Note this would not be a backwards incompatible change. We would be making an additive change here.)
+2. To retain backwards compatibility with the fact that users don’t need to provide a value field under normal circumstances, we could continue to surface the results to the `Task`.
+
+`Step` authors should be able to:
+1. reference result paths via: `$(step.results.resultName.path)` , *NOT* `$(results.resultName.path)` since `StepAction` are a new addition (backwards compatibility is not required here).
+    a .Internally, 
+      i. `$(step.results.resultName.path) = /tekton/steps/<step-name>/results/resultName`
+`Task` authors should be able to:
+1. fetch results from underlying `Steps` using  `$(step.stepName.results.resultName)`.
+2. For backwards compatibility, they can write inlined `Steps` with results written to `$(results.resultName.path)`. These results will automatically be surfaced to the `Task` by the entrypointer.
+3. However, if they write inlined `Steps` where `Results` are written to `$(step.results.resultName.path)` they need to fetch the `Step` `Results` using `$(step.stepName.results.resultName)`.
```

Comment:
Fixed. I updated the text. It is only needed when there are potential name conflicts. E.g. if two step actions produce results with the same name then the Task author can use this syntax to attach the result to another name at the Task level.

Say, step1 produces a result called `digest`.
`step2` also produces a result called `digest` and `url`

In this case, the task author needs to distinguish between step1's `digest` and step2's. There they can use:

```
results:
  - name: digest1
    value: $(step.step1.results.digest)
  - name: digest2
    value: $(step.step2.results.digest)
```

Failure to do so will cause both the results to be automatically surfaced to the Task where the second one i.e. Step2's result will overwrite Step1's.

The result `url` from `Step2` will be automatically surfaced to the Task level. But since there is no other step in the Task producing a result called `url`, the Task author does not need to resolve anything.